File: 2024/day17.pl
:- use_module(library(clpz)).
:- use_module(library(time)).
:- use_module(library(lists), [nth0/3, reverse/2, same_length/2, length/2]).

operand(state(_, _, _, _, Program, Pointer), Operand) :-
  Operand in 0 .. 7,
  OperandPointer #= Pointer + 1,
  length(Program, Len),
  OperandPointer in 0 .. Len,
  nth0(OperandPointer, Program, Operand).

combo_operand(State, Operand) :-
  Operand in 0 .. 3,
  operand(State, Operand).
combo_operand(State, RegA) :-
  operand(State, 4),
  regA(RegA, _, State, _).
combo_operand(State, RegB) :-
  operand(State, 5),
  regB(RegB, _, State, _).
combo_operand(State, RegC) :-
  operand(State, 6),
  regC(RegC, _, State, _).   

initial_state(RegA, RegB, RegC, Program, state(RegA, RegB, RegC, [], Program, 0)).

program(
  Program,
  state(_, _, _, _, Program, _)
).
regA(
  RegA,
  NewRegA,
  state(RegA, RegB, RegC, Output, Program, Pointer),
  state(NewRegA, RegB, RegC, Output, Program, Pointer)
) :- RegA #>= 0, NewRegA #>= 0.
regB(
  RegB,
  NewRegB,
  state(RegA, RegB, RegC, Output, Program, Pointer),
  state(RegA, NewRegB, RegC, Output, Program, Pointer)
) :- RegB #>= 0, NewRegB #>= 0.
regC(
  RegC,
  NewRegC,
  state(RegA, RegB, RegC, Output, Program, Pointer),
  state(RegA, RegB, NewRegC, Output, Program, Pointer)
) :- RegC #>= 0, NewRegC #>= 0.
pointer(
  Pointer,
  NewPointer,
  state(RegA, RegB, RegC, Output, Program, Pointer),
  state(RegA, RegB, RegC, Output, Program, NewPointer)   
).
output(
  Output,
  NewOutput,
  state(RegA, RegB, RegC, Output, Program, Pointer),
  state(RegA, RegB, RegC, NewOutput, Program, Pointer)   
).

inc_pointer(State0, State) :-
  pointer(Pointer, NewPointer, State0, State),
  program(Program, State0),
  length(Program, Len),
  NewPointer in 0 .. Len,
  NewPointer #= Pointer + 2.    

% The adv instruction (opcode 0) performs division. 
% The numerator is the value in the A register. 
% The denominator is found by raising 2 to the power of the instruction's combo operand. 
% (So, an operand of 2 would divide A by 4 (2^2); an operand of 5 would divide A by 2^B.) 
% The result of the division operation is truncated to an integer and then written to the 
% A register.
step(0, State0, State) :-
  combo_operand(State0, ComboOp),
  regA(RegA, NewRegA, State0, State1),
  NewRegA #= RegA div (2^ComboOp),
  inc_pointer(State1, State).
% The bxl instruction (opcode 1) calculates the bitwise XOR of register B and the instruction's 
% literal operand, then stores the result in register B.
step(1, State0, State) :-
  regB(RegB, NewRegB, State0, State1),
  operand(State0, LiteralOperand),
  NewRegB #= RegB xor LiteralOperand,
  inc_pointer(State1, State).
% The bst instruction (opcode 2) calculates the value of its combo operand modulo 8 (thereby 
% keeping only its lowest 3 bits), then writes that value to the B register.
step(2, State0, State) :-
  regB(_, NewRegB, State0, State1),
  combo_operand(State0, ComboOp),
  NewRegB #= ComboOp mod 8,
  inc_pointer(State1, State).
% The jnz instruction (opcode 3) does nothing if the A register is 0. However, if the A 
% register is not zero, it jumps by setting the instruction pointer to the value of its 
% literal operand; if this instruction jumps, the instruction pointer is not increased by 2 
% after this instruction.
step(3, State0, State) :-
  regA(0, _, State0, _),
  inc_pointer(State0, State).
step(3, State0, State) :-
  regA(RegA, _, State0, _),
  RegA #\= 0,
  operand(State0, LiteralOperand),
  pointer(_, LiteralOperand, State0, State).
% The bxc instruction (opcode 4) calculates the bitwise XOR of register B and register C, 
% then stores the result in register B. (For legacy reasons, this instruction reads an 
% operand but ignores it.)
step(4, State0, State) :-
  regB(RegB, NewRegB, State0, State1),
  regC(RegC, _      , State0, _     ),
  NewRegB #= RegB xor RegC,
  inc_pointer(State1, State).
% The out instruction (opcode 5) calculates the value of its combo operand modulo 8, 
% then outputs that value. (If a program outputs multiple values, they are separated by 
% commas.)
step(5, State0, State) :-
  combo_operand(State0, ComboOp),
  Res #= ComboOp mod 8,
  output(Output, [Res|Output], State0, State1),
  inc_pointer(State1, State).
% The bdv instruction (opcode 6) works exactly like the adv instruction except that the 
% result is stored in the B register. (The numerator is still read from the A register.)
step(6, State0, State) :-
  combo_operand(State0, ComboOp),
  regA(RegA, _, State0, _),
  NewRegB #= RegA div (2^ComboOp),
  regB(_, NewRegB, State0, State1),
  inc_pointer(State1, State).
% The cdv instruction (opcode 7) works exactly like the adv instruction except that the 
% result is stored in the C register. (The numerator is still read from the A register.)
step(7, State0, State) :-
  combo_operand(State0, ComboOp),
  regA(RegA, _, State0, _),
  NewRegC #= RegA div (2^ComboOp),
  regC(_, NewRegC, State0, State1),
  inc_pointer(State1, State).

run(State0, StateN) :-
  program(Program, State0),
  pointer(Pointer, _, State0, _),
  nth0(Pointer, Program, Instr),
  step(Instr, State0, State1),
  run(State1, StateN).
run(State, State).

 
% If register C contains 9, the program 2,6 would set register B to 1.
test1 :-
  initial_state(0, 0, 9, [2,6], State0),
  run(State0, StateN),
  regB(1, _, StateN, _).

% If register A contains 10, the program 5,0,5,1,5,4 would output 0,1,2.
test2 :-
  initial_state(10, 0, 0, [5,0,5,1,5,4], State0),
  run(State0, StateN),
  output(Output, _, StateN, _),
  reverse(Output, [0,1,2]).

% If register A contains 2024, the program 0,1,5,4,3,0 would output 4,2,5,6,7,7,7,7,3,1,0 and 
% leave 0 in register A.
test3 :- 
  initial_state(2024, 0, 0, [0,1,5,4,3,0], State0),
  run(State0, StateN),
  regA(0, _, StateN, _),
  output(Output, _, StateN, _),
  reverse(Output, [4,2,5,6,7,7,7,7,3,1,0]).

% If register B contains 29, the program 1,7 would set register B to 26.
test4 :- 
  initial_state(0, 29, 0, [1,7], State0),
  run(State0, StateN),
  regB(26, _, StateN, _).

% If register B contains 2024 and register C contains 43690, the program 4,0 would set 
% register B to 44354. 
test5 :-
  initial_state(0, 2024, 43690, [4,0], State0),
  run(State0, StateN),
  regB(44354, _, StateN, _).

test6(RegA) :- 
  Program = [0,3,5,4,3,0],
  initial_state(RegA, 0, 0, Program, State0),
  run(State0, StateN),
  reverse(Program, Output),
  output(Output, _, StateN, _),
  labeling([min(RegA)], [RegA]).

part1(RevOutput) :-
  initial_state(64196994, 0, 0, [2,4,1,1,7,5,1,5,4,0,0,3,5,5,3,0], State0),
  run(State0, StateN),
  output(Output, _, StateN, _),
  reverse(Output, RevOutput).

% 35184372088832 too low

part2(RegA) :-
  Program = [2,4,1,1,7,5,1,5,4,0,0,3,5,5,3,0],
  initial_state(RegA, 0, 0, [2,4,1,1,7,5,1,5,4,0,0,3,5,5,3,0], State0),
  RegA #>= 1,
  RegA #>= 35184372089164,
  run(State0, StateN),
  reverse(Program, Output),
  output(Output, _, StateN, _).
  % labeling([bisect], [RegA]).

stateN(StateN) :-
  Program = [2,4,1,1,7,5,1,5,4,0,0,3,5,5,3,0],
  reverse(Program, Output),
  length(Program, ProgramLength),
  Pointer #>= ProgramLength,
  StateN = state(_, _, _, Output, Program, Pointer).

File: 2024/Day17.hs
module Main (main) where
import Utils (Parser, integer, parseFile, converge)
import Text.Megaparsec.Char (string, newline, char)
import Text.Megaparsec (sepBy)
import Data.Bits (xor)
import Data.Sequence (Seq)
import qualified Data.Sequence as Seq
import Data.Foldable (traverse_, find, toList)
import Debug.Trace (traceShowId, traceShowM)
import Data.Maybe (maybeToList)
import Control.Monad (guard, when)
import Data.List (isPrefixOf)
import Data.Containers.ListUtils (nubOrd)

data State = State
  { registerA :: Int
  , registerB :: Int
  , registerC :: Int
  , output    :: [Int]
  , program   :: Seq Int
  , pointer   :: Int
  } deriving (Eq, Show)

initialState :: Int -> Int -> Int -> [Int] -> State
initialState regA regB regC instrs = 
  State 
    { registerA = regA
    , registerB = regB
    , registerC = regC
    , pointer   = 0
    , output    = []
    , program   = Seq.fromList instrs
    }

parser :: Parser State
parser = do
  regA <- string "Register A: " *> integer <* newline
  regB <- string "Register B: " *> integer <* newline
  regC <- string "Register C: " *> integer <* newline
  newline
  instrs <- string "Program: " *> integer `sepBy` char ','
  newline
  return $ initialState regA regB regC instrs

comboOperand :: State -> Int
comboOperand s
  -- Combo operands 0 through 3 represent literal values 0 through 3.
  | 0 <= operand s && operand s <= 3 = operand s
  -- Combo operand 4 represents the value of register A.
  | operand s == 4 = s.registerA
  -- Combo operand 5 represents the value of register B.
  | operand s == 5 = s.registerB
  -- Combo operand 6 represents the value of register C.
  | operand s == 6 = s.registerC
  -- Combo operand 7 is reserved and will not appear in valid programs.
  | operand s == 7 = error "combo operand 7"
  | otherwise      = error $ "combo operand " ++ show (operand s)

incPointer :: State -> State 
incPointer state = state { pointer = state.pointer + 2 }

getInstr :: State -> Maybe Int
getInstr state = Seq.lookup state.pointer state.program

operand :: State -> Int
operand state = state.program `Seq.index` (state.pointer + 1)

step :: State -> State
step state = 
  case getInstr state of 
    -- The adv instruction (opcode 0) performs division. 
    -- The numerator is the value in the A register. 
    -- The denominator is found by raising 2 to the power of the instruction's combo operand. 
    -- (So, an operand of 2 would divide A by 4 (2^2); an operand of 5 would divide A by 2^B.) 
    -- The result of the division operation is truncated to an integer and then written to the 
    -- A register.
    Just 0 -> incPointer $ state { registerA = state.registerA `div` (2 ^ comboOperand state) }
    -- The bxl instruction (opcode 1) calculates the bitwise XOR of register B and the instruction's 
    -- literal operand, then stores the result in register B.
    Just 1 -> incPointer $ state { registerB = state.registerB `xor` operand state}
    -- The bst instruction (opcode 2) calculates the value of its combo operand modulo 8 (thereby 
    -- keeping only its lowest 3 bits), then writes that value to the B register.
    Just 2 -> incPointer $ state { registerB = comboOperand state `mod` 8 }
    -- The jnz instruction (opcode 3) does nothing if the A register is 0. However, if the A 
    -- register is not zero, it jumps by setting the instruction pointer to the value of its 
    -- literal operand; if this instruction jumps, the instruction pointer is not increased by 2 
    -- after this instruction.
    Just 3 | state.registerA == 0 -> incPointer state
           | otherwise            -> state { pointer = operand state }
    -- The bxc instruction (opcode 4) calculates the bitwise XOR of register B and register C, 
    -- then stores the result in register B. (For legacy reasons, this instruction reads an 
    -- operand but ignores it.)
    Just 4 -> incPointer $ state { registerB = state.registerB `xor` state.registerC }
    -- The out instruction (opcode 5) calculates the value of its combo operand modulo 8, 
    -- then outputs that value. (If a program outputs multiple values, they are separated by 
    -- commas.)
    Just 5 -> incPointer $ state { output = (comboOperand state `mod` 8) : state.output }
    -- The bdv instruction (opcode 6) works exactly like the adv instruction except that the 
    -- result is stored in the B register. (The numerator is still read from the A register.)
    Just 6 -> incPointer $ state { registerB = state.registerA `div` (2 ^ comboOperand state) }
    -- The cdv instruction (opcode 7) works exactly like the adv instruction except that the 
    -- result is stored in the C register. (The numerator is still read from the A register.)
    Just 7 -> incPointer $ state { registerC = state.registerA `div` (2 ^ comboOperand state) }
    Just c -> error $ "undefined opcode: " ++ show c
    Nothing -> state

run :: State -> [State]
run state = 
  if (step state).pointer == state.pointer then
    [state]
  else
    state : run (step state)

search :: State -> Int
search state = head $ do
  let ints = map (2^) [0 ..] :: [Integer]
  regA <- map fromIntegral $ takeWhile (\i -> i <= fromIntegral (maxBound :: Int)) ints
  let state' = state { registerA = regA }
      p_list = toList state.program
  traceShowM regA
  let stateN = last $ takeWhile (\s -> reverse s.output `isPrefixOf` p_list) $ run state'
  when (length stateN.output > 4) (traceShowM (regA, stateN.output))
  guard $ reverse stateN.output == toList state.program
  return regA

main :: IO ()
main = do
  state0 <- parseFile parser "input/17.txt"

  putStr "Part 1: "
  print $ reverse $ output $ last $ run state0
  
  -- print $ reverse $ output $ last $ run state0
  -- traverse_ print $ run $ state0 { registerA = 8404168381 }
  -- print $ search state0
